### PR TITLE
Add XX parity gate

### PR DESCRIPTION
**Pull request by mlihc15**
Adding ion trap native 2-qubit gates.

### CODE REVIEW

**Comment by mlihc15** [comment]
Signed CLA

File: ion/ionops_test.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+# limitations under the License.
+
+import numpy as np
+import pytest
+
+import cirq
+import ion
+
+def test_CxNOTGate_init():
+    assert ion.CxNOTGate(exponent=8) == ion.CxNOTGate(exponent=0)
```
Comment (Strilanc) [inline ion/ionops_test.py]: This is testing equality, not initialization. Just check that the `.exponent` property is the value you passed in. Then add a `test_CxNotGate_eq` method using `eq = cirq.testing.EqualsTester()` and add a few equivalence groups to `eq`.

File: ion/ionops_test.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    assert np.allclose(cirq.unitary(ion.CxNOTGate()),
+                       np.array([[s, 0, 0, -1j*s],
+                                 [0, s, -1j*s, 0],
+                                 [0, -1j*s, s, 0],
+                                 [-1j*s, 0, 0, s]]))
+    assert np.allclose(cirq.unitary(ion.CxNOTGate(exponent=4)),
+                       np.diag([-1, -1, -1, -1]))
+    assert np.allclose(cirq.unitary(ion.CxNOTGate()**8),
+                       np.diag([1, 1, 1, 1]))
+
```
Comment (Strilanc) [inline ion/ionops_test.py]: Add a circuit diagram test (`c = cirq.Circuit.from_ops(cirq.CxNOT ...)` then `cirq.testing.assert_has_diagram(c, """...""")`.

File: ion/__init__.py [hunk 3]
```diff
@@ -0,0 +1,20 @@
+# Copyright 2018 The Cirq Developers
```
Comment (Strilanc) [inline ion/__init__.py]: Move to `cirq/ion`

File: ion/ionops.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+from typing import (
+    Union, Tuple, Optional, List, Callable, cast, Iterable, Sequence,
+)
+
+import numpy as np
+
+from cirq import value, protocols
+from cirq.ops import gate_features, eigen_gate
+
+class CxNOTGate(eigen_gate.EigenGate,
```
Comment (Strilanc) [inline ion/ionops.py]: Rename to `XXGate`. I guess we should also add ZZ and YY while we're at it. Can be done after.

**Comment by Strilanc** [comment]
After some conversation we realized that actually the goal is to implement an XX gate, not a CxNOT gate.

File: cirq/ops/common_gates.py [hunk 5]
```diff
@@ -982,3 +982,82 @@ def Ry(rads: float) -> RotYGate:
 def Rz(rads: float) -> RotZGate:
     """Returns a gate with the matrix e^{-i Z rads / 2}."""
     return RotZGate(rads=rads, global_shift_in_half_turns=-0.5)
+
+
+class XXGate(eigen_gate.EigenGate,
```
Comment (Strilanc) [inline cirq/ops/common_gates.py]: Rename to `XXPowGate`

File: cirq/ops/common_gates.py [hunk 6]
```diff
@@ -982,3 +982,82 @@ def Ry(rads: float) -> RotYGate:
...
+                gate_features.TwoQubitGate,
+                gate_features.InterchangeableQubitsGate):
+    """XX interaction.
+
+    The gate implements the following unitary:
+    exp(-i t X⊗X) = [cos(t) 0 0 -isin(t)]
```
Comment (Strilanc) [inline cirq/ops/common_gates.py]: The matrix of this gate should be `(X⊗X)^t = (-i)^t exp(-i pi t X⊗X)`. The current specified matrix is off by a global phase, and has a different scaling of `t`.

File: cirq/ops/common_gates.py [hunk 7]
```diff
@@ -982,3 +982,82 @@ def Ry(rads: float) -> RotYGate:
...
+                    [0 cos(t) -isin(t) 0]
+                    [0 -isin(t) cos(t) 0]
+                    [-isin(t) 0 0 cos(t)]
+
+    """
+
+    def __init__(self, *,  # Forces keyword args.
+                 duration: Optional[float]=None,
```
Comment (Strilanc) [inline cirq/ops/common_gates.py]: This duration concept is sort of unique to openfermion-cirq and has no equivalent in cirq, so we should leave it out. If we want a gate specified with alternate units, we'll make a utility method analogous to `Rx`.

Is there a typical name for a function that takes radians and spits out exp(i XX t)?

File: cirq/ops/common_gates.py [hunk 8]
```diff
@@ -982,3 +982,82 @@ def Ry(rads: float) -> RotYGate:
...
+                            [0, -0.5, 0.5, 0],
+                            [-0.5, 0, 0, 0.5]])),
+            (-0.25, np.array([[0.5, 0, 0, 0.5],
+                            [0, 0.5, 0.5, 0],
+                            [0, 0.5, 0.5, 0],
+                            [0.5, 0, 0, 0.5]]))
+        ]
+
+    def _canonical_exponent_period(self) -> Optional[float]:
+        return 8
```
Comment (Strilanc) [inline cirq/ops/common_gates.py]: The period of `f(t) = XX**t` is 2.

File: cirq/ops/common_gates.py [hunk 9]
```diff
@@ -982,3 +982,82 @@ def Ry(rads: float) -> RotYGate:
...
+
+        super().__init__(exponent=power)
+
+    @property
+    def exponent(self) -> Union[value.Symbol, float]:
+        return self._exponent
+
+    def _eigen_components(self):
+        return [
+            (0.25, np.array([[0.5, 0, 0, -0.5],
```
Comment (Strilanc) [inline cirq/ops/common_gates.py]: Use 0 and 1 instead of 0.25 and -0.25

**Comment by Strilanc** [comment]
Define the `MS(rads)` method inside of `cirq.ion`, but export it all the way to the top level.

**Pull request closed without merging**